### PR TITLE
CP-25517: Extend get_capabilities interface to show SR-IOV capability

### PR DESCRIPTION
**Pull request by cheng-z**
Signed-off-by: Cheng Zhang <cheng.zhang@citrix.com>

### CODE REVIEW

File: lib/network_utils.ml [hunk 1]
```diff
@@ -670,6 +671,19 @@ module Fcoe = struct
...
+module SRIOV = struct
+    let call ?(log=false) args =
+        call_script ~log_successful_output:log ~timeout:(Some 10.0) !sriov_plugin args
+                
+    let get_capabilities name =
+        try
+            let output = call ["--capable"; name] in
+            if String.equal (String.trim output) "1" then ["sriov"] else []
+        with _ ->
+            debug "Failed to get SRIOV support status on device %s" name;
```
Comment (krizex) [inline lib/network_utils.ml]: Please use `error` instead of `debug`

File: lib/network_utils.ml [hunk 2]
```diff
@@ -670,6 +671,19 @@ module Fcoe = struct
...
 end
 
+module SRIOV = struct
+    let call ?(log=false) args =
+        call_script ~log_successful_output:log ~timeout:(Some 10.0) !sriov_plugin args
+                
+    let get_capabilities name =
+        try
+            let output = call ["--capable"; name] in
+            if String.equal (String.trim output) "1" then ["sriov"] else []
```
Comment (krizex) [inline lib/network_utils.ml]: Make it easier to understand by:
```
if String.trim output = "1" then ["sriov"] else []
```

File: lib/network_utils.ml (hunk 1 above)
Comment (cheng-z) [reply lib/network_utils.ml]: Thanks, will change

File: lib/network_utils.ml (hunk 2 above)
Comment (cheng-z) [reply lib/network_utils.ml]: Thanks, will change

**Comment by cheng-z** [comment]
Update. @krizex . Please help to have a look again.
Appreciate your support.

File: networkd/network_server.ml [hunk 3]
```diff
@@ -373,7 +373,7 @@ module Interface = struct
 
 	let get_capabilities _ dbg ~name =
 		Debug.with_thread_associated dbg (fun () ->
-			Fcoe.get_capabilities name
+			(Fcoe.get_capabilities name) @ (SRIOV.get_capabilities name)
```
Comment (lindig) [inline networkd/network_server.ml]: I would recommend to use
```
List.concat
[ Fcoe.get_capabilities name
; SRIOV.get_capabilities name
]
```

This linear order makes it easier to add more such calls in the future.

Comment (cheng-z) [reply networkd/network_server.ml]: I agree. Thanks. Will change

File: lib/network_utils.ml [hunk 4]
```diff
@@ -670,6 +671,19 @@ module Fcoe = struct
...
+module SRIOV = struct
+    let call ?(log=false) args =
+        call_script ~log_successful_output:log ~timeout:(Some 10.0) !sriov_plugin args
+                
+    let get_capabilities name =
+        try
+            let output = call ["--capable"; name] in
+            if String.trim output = "1" then ["sriov"] else []
+        with _ ->
+            error "Failed to get SRIOV support status on device %s" name;
```
Comment (cheng-z) [inline lib/network_utils.ml]: error -> debug

File: networkd/networkd.ml [hunk 5]
```diff
@@ -48,6 +48,12 @@ let resources = [
     path = Network_utils.fcoedriver;
     perms = [ Unix.X_OK ];
   };
+  { Xcp_service.name = "sriov_plugin";
```
Comment (cheng-z) [inline networkd/networkd.ml]: _ -> -

File: lib/network_utils.ml (hunk 4 above)
Comment (lindig) [reply lib/network_utils.ml]: Other things to put into the error message:

* the string of the exception
* the location of this: `__LOC__` - this makes debugging a lot easier

**Comment by cheng-z** [comment]
Hi @robhoes 
Please help to have a look again and help to merge if no other concern.
Thanks.

File: lib/network_utils.ml [hunk 6]
```diff
@@ -35,6 +35,7 @@ let modprobe = "/sbin/modprobe"
 let ethtool = ref "/sbin/ethtool"
 let bonding_dir = "/proc/net/bonding/"
 let fcoedriver = ref "/opt/xensource/libexec/fcoe_driver"
+let sriov_plugin = ref "/opt/xensource/libexec/network_sriov"
```
Comment (minli1) [inline lib/network_utils.ml]: This should be modified due to the latest design. @cheng-z @krizex

**Pull request merged**